templates/dataset_dag/dagbegin.cmd

```diff
@@ -43,6 +43,10 @@ notify_user = {{email_to}}
 {{lines|join("\n+")}}
 requirements  = target.machine =!= MachineAttrMachine1 && target.machine =!= MachineAttrMachine2  && (isUndefined(DesiredOS) || stringListsIntersect(toUpper(DesiredOS),IFOS_installed)) && (stringListsIntersect(toUpper(target.HAS_usage_model), toUpper(my.DESIRED_usage_model))) {%if append_condor_requirements %} && {{append_condor_requirements}} {%endif%}
 
+{% if singularity_image is defined and no_singularity is false %}
```

Comment:
Right, so that's the `no_singularity` flag, but `singularity_image` will still be set to the default. I'm saying if we end up in a situation where `sigularity_image` is not set (regardless of flags), it's a bug we shouldn't ignore (by not running the job in singularity).

```py
>>> import argparse
>>> parser = argparse.ArgumentParser()
>>> group = parser.add_mutually_exclusive_group()
>>> group.add_argument('--foo', default='foo')
>>> group.add_argument('--no-foo', action='store_true')
>>> parser.parse_args([])
Namespace(foo='foo', no_foo=False)
>>> parser.parse_args(['--no-foo'])
Namespace(foo='foo', no_foo=True)

Reply:
My understanding here of your comment is this:

If `--singularity_image` is passed on the command line or no argument is given, `singularity_image` is defined (default or specified), and thus `singularity_image` will be defined, and `no_singularity` will be `false`.  Thus, we'll be inserting the `SingularityImage` line in the template. 

If `--no-singularity` is passed, then `singularity_image` will be defined (as the default), AND `no_singularity` will be `true`.  Thus, in the current conditional in the template, the `SingularityImage` line is omitted.  

That is, there is no case where `singularity_image` isn't defined, which the template currently handles correctly.  My understanding of your objection is that you're saying that one of two things should happen:

1. We should not allow `singularity_image` to be defined and `no_singularity`to be `true` at the same time.  OR
2. We should make sure the template is bulletproof, and explicitly have a clause that says that if `no_singularity` is `true`, do NOT insert the `SingularityImage` line under any circumstances.

Is that right?

Reply:
Well crap, jinja default behavior is to sub an empty string?! No I don't like that... https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.Undefined

Reply:
So changing the default behavior for undefined variables breaks a bunch of jobsub_submit unit tests, as well as some behavior with the submit, so I want to take some more time to implement that.  I think I've got it mostly done, but I don't want to rush that part to get this requested change in.  

I'll split off the default StrictUndefined stuff into another issue, PR, etc.  Is it okay to call this resolved if I split it like that?

Reply:
The commit that fixes the main objection raised is [3a3220f](https://github.com/marcmengel/jobsub_lite/pull/56/commits/3a3220fae6a3ffaddc4264dfc7b6b6961e27eff0)

Reply:
I kinda feared that. Works for me.